libc/src/math/generic/powf.cpp

```diff
@@ -0,0 +1,840 @@
+  FloatBits xbits(x), ybits(y);
+
+  uint32_t x_u = xbits.uintval();
+  uint32_t x_abs = x_u & FloatProp::EXP_MANT_MASK;
+  uint32_t y_u = ybits.uintval();
+  uint32_t y_abs = y_u & FloatProp::EXP_MANT_MASK;
+
+  ///////// BEGIN - Check exceptional cases ////////////////////////////////////
+
+  // The single precision number that is closest to 1 is (1 - 2^-24), which has
+  //   log2(1 - 2^-24) ~ -1.715...p-24.
+  // So if |y| > 151 * 2^24, and x is finite:
+  //   |y * log2(x)| = 0 or > 151.
+  // Hence x^y will either overflow or underflow if x is not zero.
+  if (LIBC_UNLIKELY((y_abs & 0x007f'ffff) == 0) || (y_abs > 0x4f170000)) {
+    // Exceptional exponents.
+    switch (y_abs) {
+    case 0x0000'0000: { // y = +-0.0f
+      return 1.0f;
+    }
+    case 0x7f80'0000: { // y = +-Inf
+      if (x_abs > 0x7f80'0000) {
+        // pow(NaN, +-Inf) = NaN
+        return x;
+      }
+      if (x_abs == 0x3f80'0000) {
+        // pow(+-1, +-Inf) = 1.0f
+        return 1.0f;
+      }
+      if (x_abs == 0 && y_u == 0xff80'0000) {
+        // pow(+-0, -Inf) = +inf and raise FE_DIVBYZERO
+        fputil::set_errno_if_required(EDOM);
+        fputil::raise_except_if_required(FE_DIVBYZERO);
+        return FloatBits::inf().get_val();
+      }
+      // pow (|x| < 1, -inf) = +inf
+      // pow (|x| < 1, +inf) = 0.0f
+      // pow (|x| > 1, -inf) = 0.0f
+      // pow (|x| > 1, +inf) = +inf
+      return ((x_abs < 0x3f80'0000) == (y_u == 0xff80'0000))
+                 ? FloatBits::inf().get_val()
+                 : 0.0f;
+    }
+    default:
+      // Speed up for common exponents
+      switch (y_u) {
+      case 0x3f00'0000: // y = 0.5f
+        // pow(x, 1/2) = sqrt(x)
+        return fputil::sqrt(x);
+      case 0x3f80'0000: // y = 1.0f
+        return x;
+      case 0x4000'0000: // y = 2.0f
+        // pow(x, 2) = x^2
+        return x * x;
+        // TODO: Enable special case speed-up for x^(-1/2) when rsqrt is ready.
+        // case 0xbf00'0000:  // pow(x, -1/2) = rsqrt(x)
+        //   return rsqrt(x);
+      }
+      if (y_abs > 0x4f17'0000) {
+        if (y_abs > 0x7f80'0000) {
+          // y is NaN
+          if (x_u == 0x3f80'0000) { // x = 1.0f
+            // pow(1, NaN) = 1
+            return 1.0f;
+          }
+          // pow(x, NaN) = NaN
+          return y;
+        }
+        // x^y will be overflow / underflow in single precision.  Set y to a
+        // large enough exponent but not too large, so that the computations
+        // won't be overflow in double precision.
+        y = cpp::bit_cast<float>((y_u & FloatProp::SIGN_MASK) + 0x4f800000U);
+      }
+    }
+  }
+
+  int ex = -FloatBits::EXPONENT_BIAS;
+  uint64_t sign = 0;
+
+  // y is finite and non-zero.
+  if (LIBC_UNLIKELY(((x_u & 0x801f'ffffU) == 0) || x_u >= 0x7f80'0000U ||
+                    x_u < 0x0080'0000U)) {
+    switch (x_u) {
+    case 0x3f80'0000: // x = 1.0f
+      return 1.0f;
+    // TODO: Put these 2 entrypoint dependency under control flag.
```

Comment:
The usual way to do this is to move the functionality into a shared utility, since we want to avoid depending on other entrypoints

Reply:
Looks like with the current cmake setup, `libc.src.errno.errno` was compiled differently between `object_library` and `entrypoint_object`.  That forces any codes using `libc_errno` to be either `header_library` or `entrypoint_object`.  I move `exp2f` and `exp10f` implementations to `exp2f_impl.h` and `exp10f_impl.h` respectively.  But I will revisit the cmake setup for `object_library` so that bigger math functions implementations can be put in a common `object_library` instead of `header_library`.

Reply:
That sounds like a good plan.